File: MapboxVelocity/MapboxVelocity/Services/DirectionsService.swift
//
//  DirectionsService.swift
//  MapboxVelocity
//
//  Created by Shaikat on 4/26/26.
//

import Foundation
import CoreLocation

struct RouteResult {
    let coordinates: [CLLocationCoordinate2D]
    let distanceMeters: Double
    let durationSeconds: Double
}

final class DirectionsService: DirectionsServiceProtocol {

    private let accessToken: String = {
        Bundle.main.object(forInfoDictionaryKey: "MBXAccessToken") as? String ?? ""
    }()

    func fetchRoute(from: CLLocationCoordinate2D,
                    to: CLLocationCoordinate2D,
                    completion: @escaping (Result<RouteResult, Error>) -> Void) {

        let urlString = "https://api.mapbox.com/directions/v5/mapbox/driving/\(from.longitude),\(from.latitude);\(to.longitude),\(to.latitude)?access_token=\(accessToken)&geometries=geojson&overview=full"

        guard let url = URL(string: urlString) else {
            completion(.failure(NSError(domain: "DirectionsService", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])))
            return
        }

        URLSession.shared.dataTask(with: url) { data, _, error in
            if let error = error {
                DispatchQueue.main.async { completion(.failure(error)) }
                return
            }

            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let routes = json["routes"] as? [[String: Any]],
                  let first = routes.first,
                  let geometry = first["geometry"] as? [String: Any],
                  let coordArrays = geometry["coordinates"] as? [[Double]],
                  let distance = first["distance"] as? Double,
                  let duration = first["duration"] as? Double
            else {
                DispatchQueue.main.async {
                    completion(.failure(NSError(domain: "DirectionsService", code: 1, userInfo: [NSLocalizedDescriptionKey: "No route found"])))
                }
                return
            }

            // Mapbox returns [longitude, latitude] — same as Geocoding
            let coordinates = coordArrays.map {
                CLLocationCoordinate2D(latitude: $0[1],
                                       longitude: $0[0])
            }

            let result = RouteResult(coordinates: coordinates,
                                     distanceMeters: distance,
                                     durationSeconds: duration)

            DispatchQueue.main.async { completion(.success(result)) }
        }.resume()
    }
}

File: MapboxVelocityAndroid/app/src/main/java/com/mapboxvelocity/Services/DirectionsService.kt
package com.mapboxvelocity.Services

import com.mapbox.geojson.Point
import com.mapboxvelocity.BuildConfig
import com.mapboxvelocity.Core.Protocols.DirectionsServiceProtocol
import com.mapboxvelocity.Core.Protocols.RouteResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject

class DirectionsService(
    private val httpClient: OkHttpClient
) : DirectionsServiceProtocol {

    private val accessToken = BuildConfig.MAPBOX_ACCESS_TOKEN

    override suspend fun fetchRoute(from: Point, to: Point): Result<RouteResult> = withContext(Dispatchers.IO) {
        val url = "https://api.mapbox.com/directions/v5/mapbox/driving/" +
                "${from.longitude()},${from.latitude()};" +
                "${to.longitude()},${to.latitude()}" +
                "?access_token=$accessToken&geometries=geojson&overview=full&steps=false"

        val response = runCatching {
            httpClient.newCall(Request.Builder().url(url).build()).execute()
        }.getOrElse { return@withContext Result.failure(it) }

        val body = response.body?.string()
            ?: return@withContext Result.failure(Exception("Empty response"))

        runCatching {
            val json = JSONObject(body)
            val route = json.getJSONArray("routes").getJSONObject(0)
            val distance = route.getDouble("distance")
            val duration = route.getDouble("duration")
            val coordsArray = route.getJSONObject("geometry").getJSONArray("coordinates")

            val coordinates = buildList {
                for (i in 0 until coordsArray.length()) {
                    val c = coordsArray.getJSONArray(i)
                    add(Point.fromLngLat(c.getDouble(0), c.getDouble(1)))
                }
            }

            RouteResult(
                coordinates = coordinates,
                distanceMeters = distance,
                durationSeconds = duration
            )
        }
    }
}
